### ISSUE TITLE
Create UTs to control xpath-suppression model regression

### ISSUE THREAD

**Issue by MEZk** [checkstyle/checkstyle#5648]
Based on https://github.com/checkstyle/checkstyle/pull/5464#issuecomment-362374643, https://github.com/checkstyle/checkstyle/pull/5464#issuecomment-362500269, https://github.com/checkstyle/checkstyle/pull/5464#issuecomment-363918259, https://github.com/checkstyle/checkstyle/issues/5508, and https://github.com/checkstyle/checkstyle/pull/5509#issuecomment-363978899, it will be good to have UTs which control similar regressions. As a starting point [this](https://github.com/checkstyle/checkstyle/pull/5509#issuecomment-363921406) algorithm can be taken into account.

@romani's suggestion
> I strongly recommend to move all that validation to UT, this will skip painful manual executions , reconfiguration, report creation. It is better to invest time ones for Test.java and config (config could build in runtime) for each Check we support, run validation from UT, run suppressions generation, verify it content by assert, append suppression to config, run one time, verify no violations. So as UTs pass, reviewer see suppressions in code and see that suppressions works.

**Comment by rnveach** [comment]
I can't speak to what your testing, but it looks like you have to do manual regression on every check? I assume this will also include any and all `log` statements inside a Check? Missing one could cause XPath to not work in that specific situation if it was coded incorrectly? Will we need every check tested if we just limit `AbstractCheck` to only `log(DetialAST, ....)` in the end?

Since this is regression of a specific check, shouldn't this stuff be in the Check's UT instead of a separate class?

**Comment by romani** [comment]
@timurt , I did comments, but please do not hesitate to make PR even for discussion. It is very good to all see my comments and cooperate on better form of test.


=========

@rnveach ,

> but it looks like you have to do manual regression on every check?

yes, it is unfortunate side effect for continuous migration from old logging(by line,column) to new(by ast).

>  I assume this will also include any and all log statements inside a Check?

we at least validate that some works, not all.

> Missing one could cause XPath to not work in that specific situation if it was coded incorrectly?

yes. Better approach is completely forbid usage of old log method. We can create new abstract class AbstractAstCheck , override all `log(line,number)` to throw exception to make it clear what should be used, and only logging by AST will be allowed. Such class will be internal (non-api), one all our Checks migrate we will think about to keep new abstract class or apply such changes(breaking compatibility) to current AbstractCheck class.
 
>  Will we need every check tested if we just limit AbstractCheck to only log(DetialAST, ....) in the end?

not all, only Checks of TreeWalker, but eventually yes.

> Since this is regression of a specific check, shouldn't this stuff be in the Check's UT instead of a separate class?

we still need to have some tests, Filters are completely standalone from Checks. Some Checks(as fileset checks will never support Xpath). I still think about Xpath ability to filter a bit independent from Checks. It will be very good in each test method of Check to have verification that all violation could be suppressed, but this will demand to update all tests to verify that all generated violation are suppres-able. BUT Xpath is only one from many filter we support. I want to postpone binding them for now, and let see how it end up (full migration, the might more nuances come up).

**Comment by MEZk** [comment]
@romani 
> BUT Xpath is only one from many filter we support. I want to postpone binding them for now, and let see how it end up (full migration, the might more nuances come up).

Agree. For now the main idea is to prevent regression. We have Xpath Suppression model which works correctly for at least some of Checkstyle checks. We should keep it working when we work on https://github.com/checkstyle/checkstyle/issues/4830.

>  Better approach is completely forbid usage of old log method.

I think that this will be the final step, but only after https://github.com/checkstyle/checkstyle/issues/4830 .

@rnveach 
> Missing one could cause XPath to not work in that specific situation if it was coded incorrectly

Unfortunately yes, because we need lineNo, tokenType, and columnNo of the violated token to generate the XPath. Thus, in the end all Checks should be changed to log AST object.

@timurt 
Please, fix issues reported by @romani and me, and send PR. The approach that you chose is OK, so lets merge the changes and move on to https://github.com/checkstyle/checkstyle/issues/4830.

**Comment by rnveach** [comment]
@romani @timurt The basic idea is to verify any AST given to `log` can be suppressed via xpath. In the manual tests, we are just using the filter to examine the AST structure against a text xpath to verify they match.
Why don't we build a test similar to [testTreeStructure](https:
//github.com/checkstyle/checkstyle/blob/6893affeac7285afd16f8a389c7d65023d0c327a/src/test/java/com/puppycrawl/tools/checkstyle/api/DetailASTTest.java#L263). We find every file and for every AST we generate, we use something like the CLI's xpath suppression generator to generate the text suppressions for us and verify that the xpath suppression filter matches the 2.

Unless @timurt is going to be widly random with his xpath suppression, it should be the same thing as adding his manual tests.

**Comment by romani** [comment]
@rnveach ,

>  The basic idea is to verify any AST given to log can be suppressed via xpath

this is true only for testing of Xpath engine, could be done as part of Filter UTs. 
Here we try to make integration testing of XxxxxCheck + XPathFilter + CLI(xpath generator). To make sure that smth we produce by CLI is actually works. CLI define line:column (not by some AST identificator), we mostly test our ability to convert "line:column" to AST. As each Check now have freedom to define any line and column for violation, it is struggle to make sure we can find/suppress violation.

Ones we finish migration of all Checks to put violation on AST, role of such tests will be not very big. But we do not know future, and it is already clear that there will be nuances with Checks that report problems on spaces (non AST elements).

As we have not time to complete whole project of conversion in one step, we need to make numerous small steps(continuous improvements), so we will do a bit more extra work.

**Comment by rnveach** [comment]
> .... we might place such tests to "src/it" (integration tests), as it is not single module testing. 

As stated at https://github.com/checkstyle/checkstyle/issues/5648#issuecomment-384265266 , since we are testing a specific check I think the code should be in the check's UT. Any common code to do the testing should be placed in `AbstractModuleTestSupport`.

If we are to make a new area for this testing, I think we should keep it separate from google regression checks. Maybe make `src/it/google` and `src/it/xpath`.

**Comment by romani** [comment]
>since we are testing a specific check I think the code should be in the check's UT.

It is not only testing of Check, it is also testing of filter and xpath generator. Yes, our UT are not pure UT for module only, they run whole process, but focused only on single module.
Here we have a case of multiple modules testing, so it doesn't belong to Check. Looks like to avoid further confusions we have to move it to "it/xpath". @rnveach, do you agree ?

**Comment by rnveach** [comment]
I am fine with `it/xpath` and we should move google to `it/google` so there is a clear separation.
I still see it as testing a specific check since it is all connected to the check's violation routine. Without a violation, there is nothing to suppress.

**Comment by romani** [comment]
:) looks like we have some mess in this area
![image](https://user-images.githubusercontent.com/812984/39547536-c1267068-4e0c-11e8-9eda-dd9eb08d4e98.png)
Definition of IT folders is not under [maven-failsafe-plugin](https://github.com/checkstyle/checkstyle/blob/master/pom.xml#L1081), but under [build-helper-maven-plugin](https://github.com/checkstyle/checkstyle/blob/master/pom.xml#L1025) , strange , now I do not remember what forced me to make it like this.
Lets do this in separate issue.

**Comment by MEZk** [comment]
@romani 
> we might place such tests to "src/it" (integration tests), as it is not single module testing. 
Looks like to avoid further confusions we have to move it to "it/xpath

Agree. Moving the tests to ```it/xpath``` makes sense.

**Comment by romani** [comment]
> Agree. Moving the tests to it/xpath makes sense.

@timurt , please do not be mislead, this update will be done in separate issue, for now follow  [my recommendation](https://github.com/checkstyle/checkstyle/issues/5648#issuecomment-386111791). We need your initial fix sooner.

**Comment by romani** [comment]
I think it is better to close this old issue and create new issue on `it` coverage for other Checks+Xpath.
